rust/datafusion/src/datasource/memory.rs

```diff
@@ -0,0 +1,194 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! In-memory data source for presenting a Vec<RecordBatch> as a data source that can be
+//! queried by DataFusion. This allows data to be pre-loaded into memory and then repeatedly
+//! queried without incurring additional file I/O overhead.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+
+use crate::datasource::{RecordBatchIterator, Table};
+use crate::execution::error::{ExecutionError, Result};
+
+/// In-memory table
+pub struct MemTable {
+    schema: Arc<Schema>,
+    batches: Vec<RecordBatch>,
+}
+
+impl MemTable {
+    /// Create a new in-memory table from the provided schema and record batches
+    pub fn new(schema: Arc<Schema>, batches: Vec<RecordBatch>) -> Result<Self> {
+        if batches
+            .iter()
+            .all(|batch| batch.num_columns() == schema.fields().len())
+        {
+            Ok(Self { schema, batches })
+        } else {
+            Err(ExecutionError::General(
+                "Mismatch between schema and batch column count".to_string(),
+            ))
+        }
+    }
+
+    /// Create a mem table by reading from another data source
+    pub fn load(t: &Table) -> Result<Self> {
+        let schema = t.schema();
+        let it = t.scan(&None, 1024 * 1024);
+        let mut it_mut = it.borrow_mut();
+
+        let mut data: Vec<RecordBatch> = vec![];
+
+        while let Ok(Some(batch)) = it_mut.next() {
+            data.push(batch);
+        }
+        MemTable::new(schema.clone(), data)
+    }
+}
+
+impl Table for MemTable {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.schema
+    }
+
+    fn scan(
+        &self,
+        projection: &Option<Vec<usize>>,
+        _batch_size: usize,
+    ) -> Rc<RefCell<RecordBatchIterator>> {
+        let columns: Vec<usize> = match projection {
+            Some(p) => p.clone(),
+            None => {
+                let l = self.schema.fields().len();
+                let mut v = Vec::with_capacity(l);
+                for i in 0..l {
+                    v.push(i);
+                }
+                v
+            }
+        };
+
+        let projected_schema = Arc::new(Schema::new(
+            columns
+                .iter()
+                .map(|i| self.schema.field(*i).clone())
```

Comment:
This may crash program with wrong arguments?

Reply:
Good catch. Fixed.